File: apps/ios-dev/ClaudeContext/Sources/Services/AI/MultiModelService.swift
import Foundation
import Combine

/// Unified multi-model AI service that supports Anthropic, OpenAI, Google, Mistral, Groq, OpenRouter, local, and custom endpoints.
@MainActor
class MultiModelService: ObservableObject {
    @Published var routingConfig = ModelRoutingConfig()
    @Published var usageStats: [String: ModelUsageStats] = [:]

    private var providerConfigs: [ModelProvider: ProviderConfig] = [:]
    private let session = URLSession.shared

    struct ProviderConfig {
        let apiKey: String
        var baseURL: String?
    }

    func configureProvider(_ provider: ModelProvider, apiKey: String, baseURL: String? = nil) {
        providerConfigs[provider] = ProviderConfig(apiKey: apiKey, baseURL: baseURL)
    }

    func setRoutingConfig(_ config: ModelRoutingConfig) {
        routingConfig = config
    }

    func getAvailableModels() -> [AiModel] {
        let enabledProviders = Set(providerConfigs.keys)
        return ModelCatalog.models.filter { enabledProviders.contains($0.provider) }
    }

    func sendMessage(
        _ messages: [ChatMessage],
        systemPrompt: String? = nil,
        modelId: String? = nil
    ) async throws -> ChatMessage {
        let model = modelId ?? routingConfig.primaryModel
        guard let aiModel = ModelCatalog.byId(model) else {
            throw AIError.modelNotFound(model)
        }
        guard let config = providerConfigs[aiModel.provider] else {
            throw AIError.providerNotConfigured(aiModel.provider.displayName)
        }

        let startTime = Date()
        let response: ChatMessage

        switch aiModel.provider {
        case .anthropic:
            response = try await sendAnthropic(messages, systemPrompt: systemPrompt, model: model, apiKey: config.apiKey)
        case .google:
            response = try await sendGoogle(messages, systemPrompt: systemPrompt, model: model, apiKey: config.apiKey)
        case .openai, .mistral, .groq, .openRouter, .local, .custom:
            let baseURL = config.baseURL ?? aiModel.provider.baseURL
            response = try await sendOpenAICompatible(messages, systemPrompt: systemPrompt, model: model, apiKey: config.apiKey, baseURL: baseURL)
        }

        // Track usage
        let latency = Int(Date().timeIntervalSince(startTime) * 1000)
        var stats = usageStats[model] ?? ModelUsageStats(modelId: model)
        stats.requestCount += 1
        stats.avgLatencyMs = (stats.avgLatencyMs * (stats.requestCount - 1) + latency) / stats.requestCount
        usageStats[model] = stats

        return response
    }

    // MARK: - Anthropic Messages API

    private func sendAnthropic(
        _ messages: [ChatMessage],
        systemPrompt: String?,
        model: String,
        apiKey: String
    ) async throws -> ChatMessage {
        let url = URL(string: "https://api.anthropic.com/v1/messages")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        request.setValue("2023-06-01", forHTTPHeaderField: "anthropic-version")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        var body: [String: Any] = [
            "model": model,
            "max_tokens": 4096,
            "messages": messages.map { msg in
                ["role": msg.role == .user ? "user" : "assistant", "content": msg.content]
            }
        ]
        if let system = systemPrompt {
            body["system"] = system
        }

        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (data, httpResponse) = try await session.data(for: request)
        try validateResponse(httpResponse)

        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        let content = (json?["content"] as? [[String: Any]])?.first?["text"] as? String ?? ""

        return ChatMessage(role: .assistant, content: content)
    }

    // MARK: - OpenAI Compatible API

    private func sendOpenAICompatible(
        _ messages: [ChatMessage],
        systemPrompt: String?,
        model: String,
        apiKey: String,
        baseURL: String
    ) async throws -> ChatMessage {
        let url = URL(string: "\(baseURL)/chat/completions")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        var apiMessages: [[String: String]] = []
        if let system = systemPrompt {
            apiMessages.append(["role": "system", "content": system])
        }
        apiMessages.append(contentsOf: messages.map { msg in
            ["role": msg.role == .user ? "user" : "assistant", "content": msg.content]
        })

        let body: [String: Any] = [
            "model": model,
            "messages": apiMessages,
            "max_tokens": 4096
        ]

        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (data, httpResponse) = try await session.data(for: request)
        try validateResponse(httpResponse)

        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        let choices = json?["choices"] as? [[String: Any]]
        let message = choices?.first?["message"] as? [String: Any]
        let content = message?["content"] as? String ?? ""

        return ChatMessage(role: .assistant, content: content)
    }

    // MARK: - Google Gemini API

    private func sendGoogle(
        _ messages: [ChatMessage],
        systemPrompt: String?,
        model: String,
        apiKey: String
    ) async throws -> ChatMessage {
        let url = URL(string: "https://generativelanguage.googleapis.com/v1beta/models/\(model):generateContent?key=\(apiKey)")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        var contents: [[String: Any]] = messages.map { msg in
            [
                "role": msg.role == .user ? "user" : "model",
                "parts": [["text": msg.content]]
            ]
        }

        var body: [String: Any] = ["contents": contents]
        if let system = systemPrompt {
            body["systemInstruction"] = ["parts": [["text": system]]]
        }

        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (data, httpResponse) = try await session.data(for: request)
        try validateResponse(httpResponse)

        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        let candidates = json?["candidates"] as? [[String: Any]]
        let content = candidates?.first?["content"] as? [String: Any]
        let parts = content?["parts"] as? [[String: Any]]
        let text = parts?.first?["text"] as? String ?? ""

        return ChatMessage(role: .assistant, content: text)
    }

    // MARK: - Helpers

    private func validateResponse(_ response: URLResponse) throws {
        guard let httpResponse = response as? HTTPURLResponse else {
            throw AIError.invalidResponse
        }
        guard (200...299).contains(httpResponse.statusCode) else {
            throw AIError.httpError(httpResponse.statusCode)
        }
    }
}

enum AIError: LocalizedError {
    case modelNotFound(String)
    case providerNotConfigured(String)
    case invalidResponse
    case httpError(Int)
    case networkError(Error)

    var errorDescription: String? {
        switch self {
        case .modelNotFound(let id): return "Model not found: \(id)"
        case .providerNotConfigured(let name): return "Provider not configured: \(name)"
        case .invalidResponse: return "Invalid response from server"
        case .httpError(let code): return "HTTP error: \(code)"
        case .networkError(let error): return "Network error: \(error.localizedDescription)"
        }
    }
}

File: apps/android-dev/app/src/main/java/com/claudecontext/localdev/service/ai/MultiModelService.kt
package com.claudecontext.localdev.service.ai

import com.claudecontext.localdev.data.models.*
import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Unified multi-model AI service that works with Anthropic, OpenAI, Google, Mistral,
 * Groq, OpenRouter, local models, and custom endpoints.
 * Provides a single sendMessage() interface regardless of the underlying provider.
 */
@Singleton
class MultiModelService @Inject constructor(
    private val gson: Gson
) {
    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(120, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()

    private val _providerConfigs = MutableStateFlow<Map<ModelProvider, ProviderConfig>>(emptyMap())
    val providerConfigs: StateFlow<Map<ModelProvider, ProviderConfig>> = _providerConfigs.asStateFlow()

    private val _usageStats = MutableStateFlow<Map<String, ModelUsageStats>>(emptyMap())
    val usageStats: StateFlow<Map<String, ModelUsageStats>> = _usageStats.asStateFlow()

    private val _routingConfig = MutableStateFlow(ModelRoutingConfig())
    val routingConfig: StateFlow<ModelRoutingConfig> = _routingConfig.asStateFlow()

    fun configureProvider(provider: ModelProvider, apiKey: String, customBaseUrl: String? = null) {
        val configs = _providerConfigs.value.toMutableMap()
        configs[provider] = ProviderConfig(
            provider = provider,
            apiKey = apiKey,
            enabled = apiKey.isNotEmpty(),
            customBaseUrl = customBaseUrl
        )
        _providerConfigs.value = configs
    }

    fun setRoutingConfig(config: ModelRoutingConfig) {
        _routingConfig.value = config
    }

    fun getConfiguredProviders(): List<ModelProvider> =
        _providerConfigs.value.filter { it.value.enabled }.keys.toList()

    fun getAvailableModels(): List<AiModel> {
        val configured = getConfiguredProviders()
        return ModelCatalog.models.filter { it.provider in configured } +
            getLocalModels()
    }

    private fun getLocalModels(): List<AiModel> {
        val localConfig = _providerConfigs.value[ModelProvider.LOCAL]
        if (localConfig?.enabled != true) return emptyList()
        return listOf(
            AiModel(
                id = "local-default",
                provider = ModelProvider.LOCAL,
                displayName = "Local Model",
                contextWindow = 8192,
                maxOutputTokens = 2048,
                isLocal = true,
                tier = ModelTier.LOCAL
            )
        )
    }

    /**
     * Send a message to any configured model.
     * Automatically adapts the request format based on the provider.
     */
    suspend fun sendMessage(
        messages: List<ClaudeMessage>,
        systemPrompt: String? = null,
        modelId: String? = null
    ): ClaudeMessage {
        val resolvedModelId = modelId ?: _routingConfig.value.primaryModel
        val model = ModelCatalog.getById(resolvedModelId)
            ?: getLocalModels().find { it.id == resolvedModelId }
            ?: throw IllegalArgumentException("Unknown model: $resolvedModelId")

        val config = _providerConfigs.value[model.provider]
            ?: throw IllegalStateException("Provider ${model.provider.displayName} not configured")

        val startTime = System.currentTimeMillis()

        val result = when (model.provider) {
            ModelProvider.ANTHROPIC -> sendAnthropic(messages, systemPrompt, model, config)
            ModelProvider.OPENAI, ModelProvider.GROQ, ModelProvider.OPENROUTER, ModelProvider.MISTRAL ->
                sendOpenAICompatible(messages, systemPrompt, model, config)
            ModelProvider.GOOGLE -> sendGoogle(messages, systemPrompt, model, config)
            ModelProvider.LOCAL -> sendOpenAICompatible(messages, systemPrompt, model, config)
            ModelProvider.CUSTOM -> sendOpenAICompatible(messages, systemPrompt, model, config)
        }

        // Track usage
        val latency = System.currentTimeMillis() - startTime
        updateUsageStats(resolvedModelId, latency)

        return result
    }

    // --- Anthropic (Claude) ---

    private suspend fun sendAnthropic(
        messages: List<ClaudeMessage>,
        systemPrompt: String?,
        model: AiModel,
        config: ProviderConfig
    ): ClaudeMessage = withContext(Dispatchers.IO) {
        val baseUrl = config.customBaseUrl ?: model.provider.baseUrl

        val requestBody = AnthropicRequest(
            model = model.id,
            maxTokens = model.maxOutputTokens,
            system = systemPrompt,
            messages = messages.filter { it.role != MessageRole.SYSTEM }.map { msg ->
                SimpleMessage(
                    role = if (msg.role == MessageRole.USER) "user" else "assistant",
                    content = msg.content
                )
            }
        )

        val request = Request.Builder()
            .url("$baseUrl/messages")
            .addHeader("x-api-key", config.apiKey)
            .addHeader("anthropic-version", "2023-06-01")
            .addHeader("content-type", "application/json")
            .post(gson.toJson(requestBody).toRequestBody("application/json".toMediaType()))
            .build()

        val response = client.newCall(request).execute()
        val body = response.body?.string() ?: throw Exception("Empty response")

        if (!response.isSuccessful) {
            throw Exception("Anthropic API error (${response.code}): $body")
        }

        val apiResponse = gson.fromJson(body, AnthropicResponse::class.java)
        val content = apiResponse.content.firstOrNull()?.text ?: ""
        ClaudeMessage(role = MessageRole.ASSISTANT, content = content, codeBlocks = extractCodeBlocks(content))
    }

    // --- OpenAI-compatible (GPT, Mistral, Groq, OpenRouter, Local) ---

    private suspend fun sendOpenAICompatible(
        messages: List<ClaudeMessage>,
        systemPrompt: String?,
        model: AiModel,
        config: ProviderConfig
    ): ClaudeMessage = withContext(Dispatchers.IO) {
        val baseUrl = config.customBaseUrl ?: model.provider.baseUrl

        val allMessages = mutableListOf<SimpleMessage>()
        if (systemPrompt != null) {
            allMessages.add(SimpleMessage(role = "system", content = systemPrompt))
        }
        allMessages.addAll(
            messages.filter { it.role != MessageRole.SYSTEM }.map { msg ->
                SimpleMessage(
                    role = if (msg.role == MessageRole.USER) "user" else "assistant",
                    content = msg.content
                )
            }
        )

        val requestBody = OpenAIRequest(
            model = model.id,
            messages = allMessages,
            maxTokens = model.maxOutputTokens
        )

        val requestBuilder = Request.Builder()
            .url("$baseUrl/chat/completions")
            .addHeader("content-type", "application/json")
            .post(gson.toJson(requestBody).toRequestBody("application/json".toMediaType()))

        // Add auth header based on provider
        if (config.apiKey.isNotEmpty()) {
            when (model.provider) {
                ModelProvider.OPENAI, ModelProvider.GROQ, ModelProvider.OPENROUTER,
                ModelProvider.MISTRAL, ModelProvider.CUSTOM ->
                    requestBuilder.addHeader("Authorization", "Bearer ${config.apiKey}")
                else -> requestBuilder.addHeader(model.provider.authHeader, config.apiKey)
            }
        }

        if (model.provider == ModelProvider.OPENROUTER) {
            requestBuilder.addHeader("HTTP-Referer", "https://claudecontext.com")
            requestBuilder.addHeader("X-Title", "Claude Local Dev")
        }

        val response = client.newCall(requestBuilder.build()).execute()
        val body = response.body?.string() ?: throw Exception("Empty response")

        if (!response.isSuccessful) {
            throw Exception("${model.provider.displayName} API error (${response.code}): $body")
        }

        val apiResponse = gson.fromJson(body, OpenAIResponse::class.java)
        val content = apiResponse.choices.firstOrNull()?.message?.content ?: ""
        ClaudeMessage(role = MessageRole.ASSISTANT, content = content, codeBlocks = extractCodeBlocks(content))
    }

    // --- Google (Gemini) ---

    private suspend fun sendGoogle(
        messages: List<ClaudeMessage>,
        systemPrompt: String?,
        model: AiModel,
        config: ProviderConfig
    ): ClaudeMessage = withContext(Dispatchers.IO) {
        val baseUrl = config.customBaseUrl ?: model.provider.baseUrl

        val contents = mutableListOf<GeminiContent>()
        messages.filter { it.role != MessageRole.SYSTEM }.forEach { msg ->
            contents.add(GeminiContent(
                role = if (msg.role == MessageRole.USER) "user" else "model",
                parts = listOf(GeminiPart(text = msg.content))
            ))
        }

        val requestBody = GeminiRequest(
            contents = contents,
            systemInstruction = systemPrompt?.let {
                GeminiContent(role = "user", parts = listOf(GeminiPart(text = it)))
            },
            generationConfig = GeminiGenerationConfig(maxOutputTokens = model.maxOutputTokens)
        )

        val request = Request.Builder()
            .url("$baseUrl/models/${model.id}:generateContent?key=${config.apiKey}")
            .addHeader("content-type", "application/json")
            .post(gson.toJson(requestBody).toRequestBody("application/json".toMediaType()))
            .build()

        val response = client.newCall(request).execute()
        val body = response.body?.string() ?: throw Exception("Empty response")

        if (!response.isSuccessful) {
            throw Exception("Google API error (${response.code}): $body")
        }

        val apiResponse = gson.fromJson(body, GeminiResponse::class.java)
        val content = apiResponse.candidates?.firstOrNull()?.content?.parts?.firstOrNull()?.text ?: ""
        ClaudeMessage(role = MessageRole.ASSISTANT, content = content, codeBlocks = extractCodeBlocks(content))
    }

    // --- Helpers ---

    private fun extractCodeBlocks(text: String): List<CodeBlock> {
        val regex = Regex("""```(\w*)\n([\s\S]*?)```""")
        return regex.findAll(text).map { match ->
            CodeBlock(code = match.groupValues[2].trim(), language = match.groupValues[1])
        }.toList()
    }

    private fun updateUsageStats(modelId: String, latencyMs: Long) {
        val stats = _usageStats.value.toMutableMap()
        val existing = stats[modelId] ?: ModelUsageStats(modelId = modelId)
        stats[modelId] = existing.copy(
            requestCount = existing.requestCount + 1,
            avgLatencyMs = ((existing.avgLatencyMs * (existing.requestCount) + latencyMs) /
                (existing.requestCount + 1))
        )
        _usageStats.value = stats
    }

    // --- Request/Response DTOs ---

    // Anthropic
    data class AnthropicRequest(
        val model: String,
        @SerializedName("max_tokens") val maxTokens: Int,
        val system: String?,
        val messages: List<SimpleMessage>
    )
    data class AnthropicResponse(
        val id: String,
        val content: List<AnthropicContentBlock>,
        val model: String
    )
    data class AnthropicContentBlock(val type: String, val text: String)

    // OpenAI-compatible
    data class OpenAIRequest(
        val model: String,
        val messages: List<SimpleMessage>,
        @SerializedName("max_tokens") val maxTokens: Int
    )
    data class OpenAIResponse(
        val id: String?,
        val choices: List<OpenAIChoice>
    )
    data class OpenAIChoice(
        val index: Int,
        val message: SimpleMessage
    )

    // Google Gemini
    data class GeminiRequest(
        val contents: List<GeminiContent>,
        @SerializedName("system_instruction") val systemInstruction: GeminiContent?,
        @SerializedName("generation_config") val generationConfig: GeminiGenerationConfig
    )
    data class GeminiContent(
        val role: String,
        val parts: List<GeminiPart>
    )
    data class GeminiPart(val text: String)
    data class GeminiGenerationConfig(
        @SerializedName("max_output_tokens") val maxOutputTokens: Int
    )
    data class GeminiResponse(
        val candidates: List<GeminiCandidate>?
    )
    data class GeminiCandidate(
        val content: GeminiContent
    )

    // Shared
    data class SimpleMessage(
        val role: String,
        val content: String
    )
}
